# Communities and crime data set
## El "data set" 
- Lo que tenemos entre manos son datos socioeconómicos del año 1990 en Estados Unidos. Un total de 1994 instancias con 128 atributos, aunque veremos cuántos de esos realmente nos sirven. 
- Sabemos además que hay valores perdidos/nulos.
- Los atributos reales están ya normalizados a un rango 0.00-1.00.
- La etiqueta es el atributo "ViolentCrimesPerPop".
## El problema
Tenemos que predecir el número de crímenes violentos en proporción a la población. Para ello tenemos muchas "features" como: el porcentaje de hogares que reciben ingresos de la seguridad social, la mediana de ingresos, el presupuesto de la policia en proporción a la población, etc... 
## Elementos
- X: Atributos predictores, nos dan 126, pero tendremos que descartar alguno no predictor.
- Y: Número de crímenes violentos en proporción a la población.
- f: Dados los atributos de una comunidad, predecir el número de crímenes violentos en proporción a su población.
## Archivo de datos
El CSV que nos proporcionan no tiene el nombre de los atributos, así que lo primero que he hecho es extraer el nombre de los mismos del archivo que nos dan (communities.name) y añadirlo en el CSV para que sea más fácil preprocesarlos.


In [1]:
import pandas as pd

raw_data = pd.read_csv('./datos/communities.data')
raw_data

state county community        communityname  fold  population  \
0         8      ?         ?         Lakewoodcity     1        0.19   
1        53      ?         ?          Tukwilacity     1        0.00   
2        24      ?         ?         Aberdeentown     1        0.00   
3        34      5     81440  Willingborotownship     1        0.04   
4        42     95      6096    Bethlehemtownship     1        0.01   
...     ...    ...       ...                  ...   ...         ...   
1989     12      ?         ?    TempleTerracecity    10        0.01   
1990      6      ?         ?          Seasidecity    10        0.05   
1991      9      9     80070        Waterburytown    10        0.16   
1992     25     17     72600          Walthamcity    10        0.08   
1993      6      ?         ?          Ontariocity    10        0.20   

      householdsize  racepctblack  racePctWhite  racePctAsian  ...  LandArea  \
0              0.33          0.02          0.90          0.12  ...      0.12   
1              0.16          0.12          0.74          0.45  ...      0.02   
2              0.42          0.49          0.56          0.17  ...      0.01   
3              0.77          1.00          0.08          0.12  ...      0.02   
4              0.55          0.02          0.95          0.09  ...      0.04   
...             ...           ...           ...           ...  ...       ...   
1989           0.40          0.10          0.87          0.12  ...      0.01   
1990           0.96          0.46          0.28          0.83  ...      0.02   
1991           0.37          0.25          0.69          0.04  ...      0.08   
1992           0.51          0.06          0.87          0.22  ...      0.03   
1993           0.78          0.14          0.46          0.24  ...      0.11   

      PopDens  PctUsePubTrans  PolicCars  PolicOperBudg  LemasPctPolicOnPatr  \
0        0.26            0.20       0.06           0.04                  0.9   
1        0.12            0.45          ?              ?                    ?   
2        0.21            0.02          ?              ?                    ?   
3        0.39            0.28          ?              ?                    ?   
4        0.09            0.02          ?              ?                    ?   
...       ...             ...        ...            ...                  ...   
1989     0.28            0.05          ?              ?                    ?   
1990     0.37            0.20          ?              ?                    ?   
1991     0.32            0.18       0.08           0.06                 0.78   
1992     0.38            0.33       0.02           0.02                 0.79   
1993     0.30            0.05       0.08           0.04                 0.73   

      LemasGangUnitDeploy  LemasPctOfficDrugUn  PolicBudgPerPop  \
0                     0.5                 0.32             0.14   
1                       ?                 0.00                ?   
2                       ?                 0.00                ?   
3                       ?                 0.00                ?   
4                       ?                 0.00                ?   
...                   ...                  ...              ...   
1989                    ?                 0.00                ?   
1990                    ?                 0.00                ?   
1991                    0                 0.91             0.28   
1992                    0                 0.22             0.18   
1993                  0.5                 1.00             0.13   

      ViolentCrimesPerPop  
0                    0.20  
1                    0.67  
2                    0.43  
3                    0.12  
4                    0.03  
...                   ...  
1989                 0.09  
1990                 0.45  
1991                 0.23  
1992                 0.19  
1993                 0.48  

[1994 rows x 128 columns]

## Eliminando atributos no predictores
He considerado adecuado eliminar algunos atributos considerados no predictores:
- County: Es un código de area administrativa y además tiene muchos datos perdidos.
- Community: Un código que identifica la comunidad en cuestión, no nos interesa porque es un identificador de una instancia en concreto.
- Community Name: Lo mismo, pero con el nombre de la comunidad correspondiente al código.
- Fold: Nos proporcionan ya una descomposición para validación cruzada, la desechamos.
- State: Tal como está dado no puede usarse. Ya que confundiría a los algoritmos de entrenamiento (que un estado tenga el código 20 no significa que sea más que uno con código 3). Al principio iba a tratar de codificar este attributo categórico con un "One Hot Encoder", pero al ser tantos estados distintos el número de categorías iba a aumentar demasiado.

In [2]:
data_without_some_columns = raw_data.drop(['state','county', 'community', 'communityname', 'fold'], axis=1)
data_without_some_columns

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
...          ...            ...           ...           ...           ...   
1989        0.01           0.40          0.10          0.87          0.12   
1990        0.05           0.96          0.46          0.28          0.83   
1991        0.16           0.37          0.25          0.69          0.04   
1992        0.08           0.51          0.06          0.87          0.22   
1993        0.20           0.78          0.14          0.46          0.24   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  \
0            0.17         0.34         0.47         0.29        0.32  ...   
1            0.07         0.26         0.59         0.35        0.27  ...   
2            0.04         0.39         0.47         0.28        0.32  ...   
3            0.10         0.51         0.50         0.34        0.21  ...   
4            0.05         0.38         0.38         0.23        0.36  ...   
...           ...          ...          ...          ...         ...  ...   
1989         0.16         0.43         0.51         0.35        0.30  ...   
1990         0.32         0.69         0.86         0.73        0.14  ...   
1991         0.25         0.35         0.50         0.31        0.54  ...   
1992         0.10         0.58         0.74         0.63        0.41  ...   
1993         0.77         0.50         0.62         0.40        0.17  ...   

      LandArea  PopDens  PctUsePubTrans  PolicCars  PolicOperBudg  \
0         0.12     0.26            0.20       0.06           0.04   
1         0.02     0.12            0.45          ?              ?   
2         0.01     0.21            0.02          ?              ?   
3         0.02     0.39            0.28          ?              ?   
4         0.04     0.09            0.02          ?              ?   
...        ...      ...             ...        ...            ...   
1989      0.01     0.28            0.05          ?              ?   
1990      0.02     0.37            0.20          ?              ?   
1991      0.08     0.32            0.18       0.08           0.06   
1992      0.03     0.38            0.33       0.02           0.02   
1993      0.11     0.30            0.05       0.08           0.04   

      LemasPctPolicOnPatr  LemasGangUnitDeploy  LemasPctOfficDrugUn  \
0                     0.9                  0.5                 0.32   
1                       ?                    ?                 0.00   
2                       ?                    ?                 0.00   
3                       ?                    ?                 0.00   
4                       ?                    ?                 0.00   
...                   ...                  ...                  ...   
1989                    ?                    ?                 0.00   
1990                    ?                    ?                 0.00   
1991                 0.78                    0                 0.91   
1992                 0.79                    0                 0.22   
1993                 0.73                  0.5                 1.00   

      PolicBudgPerPop  ViolentCrimesPerPop  
0                0.14                 0.20  
1                   ?                 0.67  
2                   ?                 0.43  
3                   ?                 0.12  
4                   ?                 0.03  
...               ...                  ...  
1989                ?                 0.09  
1990                ?                 0.45  
1991             0.28                 0.23  
1992             0.18   

## ¿Qué hacemos con los datos perdidos?
Al ser un data set real, es normal que nos encontremos con datos que faltan. Vamos a echar un vistazo, a ver cuantos datos faltan en cada columna.

In [3]:
for i,c in (data_without_some_columns == '?').sum().items() :
    if c > 0:
        print(i,c)
  

OtherPerCap 1
LemasSwornFT 1675
LemasSwFTPerPop 1675
LemasSwFTFieldOps 1675
LemasSwFTFieldPerPop 1675
LemasTotalReq 1675
LemasTotReqPerPop 1675
PolicReqPerOffic 1675
PolicPerPop 1675
RacialMatchCommPol 1675
PctPolicWhite 1675
PctPolicBlack 1675
PctPolicHisp 1675
PctPolicAsian 1675
PctPolicMinor 1675
OfficAssgnDrugUnits 1675
NumKindsDrugsSeiz 1675
PolicAveOTWorked 1675
PolicCars 1675
PolicOperBudg 1675
LemasPctPolicOnPatr 1675
LemasGangUnitDeploy 1675
PolicBudgPerPop 1675


/home/alex/.local/lib/python3.6/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Podemos ver que hay unas cuantas columnas en esta situación, tenemos tres opciones:
- Eliminar las instancias que tienen datos perdidos en alguna de las categorías (la descartamos, porque son demasiadas).
- Eliminar las columnas con datos perdidos.
- De alguna manera rellenar los datos que faltan. Esta sería quizás la ideal, pero no tenemos la información suficiente para inferir los datos perdidos.

Tendremos que eliminar las columnas con datos perdidos. Salvo "OtherPerCap" que solamente tiene 1 instancia sin información. En ese caso simplemente buscamos dicha instancia y la eliminamos para que no cause problemas.

In [4]:
data_without_some_columns.index[data_without_some_columns['OtherPerCap'] == '?'].tolist()
data_without_some_columns = data_without_some_columns.drop(130)

Por último, eliminamos las columnas con 1675 datos perdidos:

In [5]:
columns_to_drop = [i for i,c in (data_without_some_columns == '?').sum().items() if c > 0]
print(columns_to_drop)
clean_data = data_without_some_columns.drop(columns_to_drop, axis=1)
clean_data

['LemasSwornFT', 'LemasSwFTPerPop', 'LemasSwFTFieldOps', 'LemasSwFTFieldPerPop', 'LemasTotalReq', 'LemasTotReqPerPop', 'PolicReqPerOffic', 'PolicPerPop', 'RacialMatchCommPol', 'PctPolicWhite', 'PctPolicBlack', 'PctPolicHisp', 'PctPolicAsian', 'PctPolicMinor', 'OfficAssgnDrugUnits', 'NumKindsDrugsSeiz', 'PolicAveOTWorked', 'PolicCars', 'PolicOperBudg', 'LemasPctPolicOnPatr', 'LemasGangUnitDeploy', 'PolicBudgPerPop']


population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
...          ...            ...           ...           ...           ...   
1989        0.01           0.40          0.10          0.87          0.12   
1990        0.05           0.96          0.46          0.28          0.83   
1991        0.16           0.37          0.25          0.69          0.04   
1992        0.08           0.51          0.06          0.87          0.22   
1993        0.20           0.78          0.14          0.46          0.24   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  \
0            0.17         0.34         0.47         0.29        0.32  ...   
1            0.07         0.26         0.59         0.35        0.27  ...   
2            0.04         0.39         0.47         0.28        0.32  ...   
3            0.10         0.51         0.50         0.34        0.21  ...   
4            0.05         0.38         0.38         0.23        0.36  ...   
...           ...          ...          ...          ...         ...  ...   
1989         0.16         0.43         0.51         0.35        0.30  ...   
1990         0.32         0.69         0.86         0.73        0.14  ...   
1991         0.25         0.35         0.50         0.31        0.54  ...   
1992         0.10         0.58         0.74         0.63        0.41  ...   
1993         0.77         0.50         0.62         0.40        0.17  ...   

      PctForeignBorn  PctBornSameState  PctSameHouse85  PctSameCity85  \
0               0.12              0.42            0.50           0.51   
1               0.21              0.50            0.34           0.60   
2               0.14              0.49            0.54           0.67   
3               0.19              0.30            0.73           0.64   
4               0.11              0.72            0.64           0.61   
...              ...               ...             ...            ...   
1989            0.22              0.28            0.34           0.48   
1990            0.53              0.25            0.17           0.10   
1991            0.25              0.68            0.61           0.79   
1992            0.45              0.64            0.54           0.59   
1993            0.68              0.50            0.34           0.35   

      PctSameState85  LandArea  PopDens  PctUsePubTrans  LemasPctOfficDrugUn  \
0               0.64      0.12     0.26            0.20                 0.32   
1               0.52      0.02     0.12            0.45                 0.00   
2               0.56      0.01     0.21            0.02                 0.00   
3               0.65      0.02     0.39            0.28                 0.00   
4               0.53      0.04     0.09            0.02                 0.00   
...              ...       ...      ...             ...                  ...   
1989            0.39      0.01     0.28            0.05                 0.00   
1990            0.00      0.02     0.37            0.20                 0.00   
1991            0.76      0.08     0.32            0.18                 0.91   
1992            0.52      0.03     0.38            0.33                 0.22   
1993            0.68      0.11     0.30            0.05                 1.00   

      ViolentCrimesPerPop  
0                    0.20  
1                    0.67  
2                    0.43  
3                    0.12  
4                    0.03  
...                   ...  
1989                 0.09  
1990                 0.45  
1991                 0.23  
1992                 0.19  
1993      

In [25]:
# Ahora que no tenemos ningún '?' podemos asegurarnos de que todas las columnas son de tipo float64
clean_data = clean_data.astype('float64')

In [36]:
X = clean_data.drop('ViolentCrimesPerPop', axis=1).to_numpy()
Y = clean_data['ViolentCrimesPerPop'].to_numpy()
print('X shape:', X.shape)
print('Y shape:', Y.shape)

X shape: (1993, 100)
Y shape: (1993,)
